MatrixInterpolator.C

```diff
@@ -0,0 +1,408 @@
+
+                Matrix* log_eigs = new Matrix(log_eigenpair.eigs.size(), log_eigenpair.eigs.size(), false);
+                for (int i = 0; i < log_eigenpair.eigs.size(); i++)
+                {
+                    CAROM_VERIFY(log_eigenpair.eigs[i] > 0);
+                    log_eigs->item(i, i) = std::log(log_eigenpair.eigs[i]);
+                }
+
+                // Invert matrix.
+                log_ev->inverse(log_ev_inv);
+
+                // Perform log mapping.
+                Matrix* log_ev_mult_log_eig = log_ev->mult(log_eigs);
+                Matrix* gamma = log_ev_mult_log_eig->mult(log_ev_inv);
+                gammas.push_back(gamma);
+
+                delete log_ev;
+                delete log_ev_inv;
+                delete log_eigs;
+                delete log_ev_mult_log_eig;
+            }
+        }
+
+        delete x_half_power_inv;
+
+        // Obtain lambda for the P interpolation matrix
+        obtainLambda(gammas);
+    }
+
+    // Obtain distances from database points to new point
+    std::vector<double> inv_q = obtainRBF(point);
+
+    // Interpolate gammas to get gamma for new point
+    Matrix* log_interpolated_matrix = obtainLogInterpolatedMatrix(inv_q);
+
+    // Diagonalize the new gamma so we can exponentiate it
+    EigenPair exp_eigenpair = SymmetricRightEigenSolve(log_interpolated_matrix);
+    delete log_interpolated_matrix;
+
+    Matrix* exp_ev = exp_eigenpair.ev;
+    Matrix* exp_ev_inv = NULL;
+
+    Matrix* exp_eigs = new Matrix(exp_eigenpair.eigs.size(), exp_eigenpair.eigs.size(), false);
+    for (int i = 0; i < exp_eigenpair.eigs.size(); i++)
+    {
+        exp_eigs->item(i, i) = std::exp(exp_eigenpair.eigs[i]);
+    }
+
+    exp_ev->inverse(exp_ev_inv);
+    Matrix* exp_ev_mult_exp_eig = exp_ev->mult(exp_eigs);
+
+    // Expoentiate gamma
+    Matrix* exp_gamma = exp_ev_mult_exp_eig->mult(exp_ev_inv);
+
+    delete exp_ev;
+    delete exp_ev_inv;
+    delete exp_eigs;
+    delete exp_ev_mult_exp_eig;
+
+    // Obtain exp mapping by doing X^1/2*exp(gamma)*X^1/2
+    Matrix* x_half_power_mult_exp_gamma = d_x_half_power->mult(exp_gamma);
+    Matrix* interpolated_matrix = x_half_power_mult_exp_gamma->mult(d_x_half_power);
+
+    delete  x_half_power_mult_exp_gamma;
+    return interpolated_matrix;
+
+}
+
+Matrix* MatrixInterpolator::interpolateNonSingularMatrix(Vector* point)
+{
+    if (d_lambda_T == NULL)
+    {
+        // Invert X
+        Matrix* ref_matrix_inv = NULL;
+        d_rotated_reduced_matrices[d_ref_point]->inverse(ref_matrix_inv);
+
+        // Perform log mapping.
+        std::vector<Matrix*> gammas;
+
+        for (int i = 0; i < d_parameter_points.size(); i++)
+        {
+            // For ref_point, gamma is the zero matrix
+            if (i == d_ref_point)
+            {
+                Matrix* gamma = new Matrix(ref_matrix_inv->numRows(), ref_matrix_inv->numColumns(), ref_matrix_inv->distributed());
+                gammas.push_back(gamma);
```

Comment:
`gamma = 0.0;`?

Reply:
The constructor of a new matrix by default sets every value to 0.0.

Reply:
I checked the constructors for `Matrix`, and it looks like it just calls `setSize`, which allocates memory without initializing. Am I missing something?

Reply:
I think we talked about this in another PR. Currently, setSize destroys the original data and creates a new empty memory.

d_mat = new double [new_size] {};

The {} actually initializes everything to zero.

Reply:
https://stackoverflow.com/questions/201101/how-to-initialize-all-members-of-an-array-to-the-same-value

Reply:
To be safe, I can change {} to {0}

Reply:
Thanks for clarifying, I did forget about this.